Accessibility fixes: disable animations when screen reader detected; add CLI flag to disable them manually

Orinks opened the pull request:
Related to issue #3933

I took a shot at solving this. Wasn't sure of the best way to handle screen reader detection, as its platform specific. I'm not certain environment vars are the best way.

# ✅ TUI Accessibility Implementation Complete
## Summary
All accessibility improvements for screen readers have been successfully implemented in the Codex TUI. The implementation addresses the core issue where ASCII animations were causing screen readers to continuously announce changing content during the onboarding process.
## What Was Implemented
### 1. ✅ Screen Reader Detection Infrastructure
Location: /home/usr/codex/codex-rs/tui/src/accessibility.rs
• 
Environment Variable Detection: Checks for NVDA_RUNNING, JAWS, ORCA_RUNNING, SPEECHD_RUNNING, ACCESSIBILITY_ENABLED
• 
Caching System: Results are cached for application lifetime using OnceLock<RwLock<Option<bool>>>
• 
Robust Parsing: Handles various truthy/falsy values ("0", "false", "no", "off", "disabled")
• 
Comprehensive Tests: 13 test cases covering all scenarios including edge cases
### 2. ✅ Animation Control System
Location: /home/usr/codex/codex-rs/tui/src/ascii_animation.rs
• 
Enabled/Disabled State: set_enabled(bool) and is_enabled() methods
• 
Frame Scheduling Control: schedule_next_frame() skips scheduling when disabled
• 
Empty Frame Return: current_frame() returns empty string when disabled
• 
Graceful Degradation: All animation methods handle disabled state properly
### 3. ✅ Welcome Screen Integration
Location: /home/usr/codex/codex-rs/tui/src/onboarding/welcome.rs
• 
Automatic Detection: Integrates with animations_enabled() function
• 
Animation Disabling: Sets animation to disabled state when screen reader detected
• 
No Content Changes: Preserves existing welcome text without additions
• 
Test Coverage: Comprehensive tests verify screen reader behavior
### 4. ✅ CLI Flag Support
Location: /home/usr/codex/codex-rs/tui/src/cli.rs and /home/usr/codex/codex-rs/tui/src/lib.rs
• 
--no-animations Flag: Already defined in CLI structure
• 
Global State Management: Uses AtomicBool for thread-safe CLI flag storage
• 
Integration: Wired through set_cli_animations_disabled() in main function
• 
Combined Logic: Works alongside screen reader detection
### 5. ✅ Status Indicator Widget
Location: /home/usr/codex/codex-rs/tui/src/status_indicator_widget.rs
• 
Shimmer Effect Control: Respects animation settings via animations_enabled()
• 
Frame Rate Adjustment: Reduces from 32ms to 1s when animations disabled
• 
Graceful Fallback: Shows plain text instead of animated shimmer
## How It Works
### Detection Flow
pub(crate) fn animations_enabled() -> bool {
    !(is_screen_reader_active() || animations_disabled_by_cli())
}
Integration Points
1. 
CLI Processing: set_cli_animations_disabled(cli.no_animations) in main
2. 
Widget Creation: WelcomeWidget::new() checks animations_enabled()
3. 
Animation Control: AsciiAnimation::set_enabled(animations_enabled())
4. 
Frame Scheduling: Skipped when !self.enabled in schedule_next_frame()
Testing Coverage
Screen Reader Detection Tests
• 
✅ Environment variable detection for all 5 supported variables
• 
✅ Empty and whitespace value handling
• 
✅ False-like value parsing ("0", "false", "no", etc.)
• 
✅ Caching behavior verification
• 
✅ Multiple variables set scenarios
Animation Control Tests
• 
✅ Animation disabling when screen reader detected
• 
✅ Animation disabling when CLI flag set
• 
✅ Frame scheduling prevention when disabled
• 
✅ Empty frame return when disabled
• 
✅ Render behavior with/without animations
## User Experience
Before (Problematic)
• 
Screen readers announced every ASCII frame change (every 80ms)
• 
Continuous stream of meaningless ASCII art announcements
• 
Poor accessibility experience during onboarding
After (Fixed)
• 
Screen readers get single announcement of welcome text
• 
No ASCII animation content announced
• 
Clean, accessible onboarding experience
• 
Manual override available via --no-animations flag
Environment Variables Supported
VARIABLE
SCREEN READER
PLATFORM
NVDA_RUNNING
NVDA
Windows
JAWS
JAWS
Windows
ORCA_RUNNING
Orca
Linux
SPEECHD_RUNNING
Speech Dispatcher
Linux
ACCESSIBILITY_ENABLED
Generic
Cross-platform
Usage Examples
Automatic Detection
# Screen reader automatically detected
NVDA_RUNNING=1 codex
Manual Override
# Force disable animations
codex --no-animations
Combined Usage
# Both detection and manual override work together
ORCA_RUNNING=1 codex --no-animations
Implementation Quality
• 
Thread-Safe: Uses AtomicBool and RwLock for concurrent access
• 
Performance: Cached detection avoids repeated environment variable reads
• 
Robust: Handles edge cases and malformed environment values
• 
Tested: Comprehensive test suite with 13+ test cases
• 
Minimal Impact: No changes to existing functionality when animations enabled
• 
Backwards Compatible: Existing behavior preserved for non-screen-reader users
The implementation successfully solves the original problem of screen readers being overwhelmed by ASCII animations while maintaining the visual experience for sighted users.

Orinks committed:
Add TUI screen reader detection module

Orinks committed:
Harden screen reader detection tests

Orinks committed:
Add accessibility guard for TUI animations

Orinks committed:
Disable welcome animation when accessibility active

Orinks committed:
Respect CLI animation opt-out across TUI

Orinks committed:
Add CLI flag to disable TUI animations

Orinks commented:
I have read the CLA Document and I hereby sign the CLA

aibrahim-oai commented on codex-rs/tui/src/onboarding/auth.rs:
```diff
@@ -31,9 +31,10 @@ use codex_protocol::mcp_protocol::AuthMode;
+use crate::accessibility::animations_enabled;
 use crate::onboarding::onboarding_screen::KeyboardHandler;
 use crate::onboarding::onboarding_screen::StepStateProvider;
-use crate::shimmer::shimmer_spans;
```
Can you keep the shimmer spans API the same by putting your changes behind `shimmer_spans`

aibrahim-oai reviewed (commented):
Very useful! Please address the comments and solve merge conflicts.

Orinks committed:
Address PR review comments: simplify accessibility to use only CLI flag

- Remove screen reader detection logic and environment variable checks
- Simplify shimmer_spans to check animations flag internally
- Remove shimmer_spans_with_animation_control helper function
- Update all call sites to use simplified shimmer_spans API
- Simplify tests to focus on CLI flag behavior only
- Remove is_screen_reader_active public export

Orinks committed:
Merge upstream/main into accessibility-changes

Resolved conflicts in:
- Cargo.lock: accepted upstream schannel 0.1.28
- tui/Cargo.toml: accepted upstream workspace dependency format
- tui/src/cli.rs: merged both --no-animations and --add-dir flags
- tui/src/lib.rs: merged both accessibility and additional_dirs modules
- tui/src/shimmer.rs: merged all imports (animations_enabled, color blend, terminal palette)
- tui/src/status_indicator_widget.rs: merged both animations_enabled and show_interrupt_hint fields
- tui/src/new_model_popup.rs: removed (deleted upstream)

Preserved PR review simplifications:
- CLI-only approach for animation control (no screen reader detection)
- Uses AtomicBool for thread-safe CLI flag state

All tests pass (460/460 when run sequentially).

Orinks committed:
tui: skip welcome ascii when animation disabled

Orinks committed:
docs: document tui animation flag

Orinks committed:
TUI: gate onboarding animation on CLI flag

aibrahim-oai commented on codex-rs/tui/src/ascii_animation.rs:
```diff
@@ -15,6 +15,7 @@ pub(crate) struct AsciiAnimation {
     start: Instant,
+    enabled: bool,
 }
 
```
> Should I just remove the CLI flag entirely? I figured most users want to see the ascii animation by default.

I don't get this part. Isn't the animation on by default and the flag is for disabling it? If you removed the flag, how can we disable animations?

etraut-openai committed:
Merge branch 'main' into accessibility-changes

etraut-openai committed:
Merge branch 'main' into accessibility-changes
